daemon/tests/test.rs

```diff
@@ -0,0 +1,396 @@
+                                   .arg("-o")
+                                   .arg("ppid=")
+                                   .arg("-p")
+                                   .arg(format!("{}", pid))
+                                   .output()?
+                                   .stdout)?;
+    let pid: libc::pid_t = output.trim().parse()
+        .with_context(|| format!("parsing 'ps' output \"{output}\""))?;
+    Ok(pid)
+}
+
+#[cfg(all(target_os = "linux", target_env = "gnu"))]
+fn check_process_name(pid: libc::pid_t, expected_name: &str) -> Result<()> {
+    let path: PathBuf = format!("/proc/{pid}/comm").into();
+    let actual_name = String::from_utf8(std::fs::read(&path)?)?;
+    assert_eq!(actual_name.trim(), expected_name);
+    Ok(())
+}
+
+#[cfg(not(all(target_os = "linux", target_env = "gnu")))]
+fn check_process_name(_pid: libc::pid_t, _expected_name: &str) -> Result<()> {
+    Ok(())
+}
+
+fn remove_if_exists<P: AsRef<Path>>(path: P) -> Result<()> {
+    match std::fs::remove_file(path) {
+        Ok(()) => Ok(()),
+        Err(e) if e.kind() == ErrorKind::NotFound => Ok(()),
+        Err(e) => Err(e)?,
+    }
+}
+
+enum Completion<T> {
+    Incomplete,
+    Complete(T)
+}
+use Completion::*;
+
+/// Repeatedly evaluates `condition`, sleeping a bit between calls, until it yields
+/// Complete(value), then returns Ok(value).  After a while, however, give up and return an error
+/// instead.
+fn wait_until<T, F>(mut condition: F) -> Result<T>
+    where F: FnMut() -> Completion<T>
+{
+    for i in 0..10 {
+        if let Complete(result) = condition() {
+            return Ok(result)
+        }
+        let ms = match i {
+            0 => 10,
+            1 => 100,
+            _ => 1000,
+        };
+        std::thread::sleep(std::time::Duration::from_millis(ms));
+    }
+    Err(anyhow!("wait_until timed out"))
+}
+
+fn test_daemon_command() -> Result<Command>
+{
+    let examples_dir = std::env::current_dir()?.join("target/debug/examples");
+    Ok(Command::new(examples_dir.join("test-daemon")))
+}
+
+fn unique_filename(extension: &str) -> Result<PathBuf> {
+    static COUNTER: Mutex<usize> = Mutex::new(0);
+    let count = match *COUNTER.lock().unwrap() {
+        ref mut counter => { *counter += 1; *counter }
+    };
+
+    let pid = std::process::id();
+    let pidfile_name: PathBuf = format!("test{pid}.{count}.{extension}").into();
+    remove_if_exists(&pidfile_name)?;
+    Ok(pidfile_name)
+}
+
+fn pidfile_name() -> Result<PathBuf> {
+    unique_filename("pid")
+}
+
+fn send_signal(pid: libc::pid_t, signal: c_int) -> Result<(), std::io::Error> {
+    if unsafe { libc::kill(pid, signal) } < 0 {
+        Err(std::io::Error::last_os_error())?
+    } else {
+        Ok(())
+    }
+}
+
+fn init_tracing() {
+    tracing_subscriber::fmt().with_writer(std::io::stderr).init();
+}
+
+fn process_exists(pid: libc::pid_t) -> Result<(), std::io::Error> {
+    send_signal(pid, 0)
```

Comment:
this is also racy... but at least you know the process existed at some point.

Reply:
It's used in a way here such that it is suitable for assertions. In general, yes, it's racy.